### PR TITLE
a2b_base64: Ignore invalid chars like CPython

### PR DESCRIPTION
**Pull request by alex-robbins**
Like CPython, MicroPython's `b2a_base64` function adds a `'\n'` to the encoded data. For *decoding*, however, CPython's `a2b_base64` ignores invalid characters in its input, while MicroPython's does not. Besides hurting compatibility with CPython, this means that MicroPython cannot decode its own Base64-encoded data; `a2b_base64(b2a_base64(b'foo'))` fails.

In this patch I completely rewrote the `a2b_base64` function to ignore invalid characters, basing it heavily on CPython's implementation. It is intended to work identically; if anyone finds any discrepancies, that'd be worth mentioning.

This change reduces code size on my 64-bit Unix platform by 160 bytes.

### CODE REVIEW

**Comment by stinos** [comment]
At first sight this is ok but it should adhere to the [codeconventions](https://github.com/micropython/micropython/blob/master/CODECONVENTIONS.md)

**Comment by alex-robbins** [comment]
You're right; somehow I missed that. Sorry.

Fixed the style issues, rebased onto master, and force pushed. All changes were superficial (whitespace, braces, comments, and commit messages).

Let me know if I missed anything.

**Review by pfalcon — CHANGES_REQUESTED** [review]
> basing it heavily on CPython's implementation

This becomes a common pitfall (please see https://github.com/micropython/micropython/commit/b8fef67c696f3fc5dfe8c4144c4a3fca1c769674). MicroPython is from-scratch implementation, a clean IP. Please don't add code "heavily based" on other code with incompatible license to it (CPython's license is incompatible with MicroPython's very simple MIT license).

Technical comments:

1.

> Decode base64-encoded data, ignoring invalid characters in the input. Conforms to `RFC 3548 <https://tools.ietf.org/html/rfc3548.html>` 

Can you please quote that document where it says that invalid data:

a) MAY be ignored without affecting decoding
b) MUST be ignored without affecting decoding

(MAY/MUST per RFC usage.)

2. Claim of decreasing size by 160 bytes is definitely a good plus for this patch, but I wonder if same (more) effect could have been achieved by eliminating common subexpressions like `in[j]`.

**Comment by alex-robbins** [comment]
#### Copyright Concerns
I did consider this after writing my original comment. "heavily based" may have been misleading. What I meant was that my implementation was designed to have the same *behavior* as CPython's, not that the implementations *themselves* are identical (and they aren't).

I certainly didn't copy and paste anything. I did refer to CPython's implementation for two reasons:
1. To see exactly how CPython would behave (which I think you said is okay in the first sentence of [this comment](https://github.com/micropython/micropython/pull/3090#issuecomment-301607729))
2. To get the idea of keeping an integer to shift bits in and out of. This is the `shift` integer in my code; the analogue in CPython's implementation would be `leftchar`.

That `shift` integer is the only idea I'm aware of getting from CPython. I basically just read the CPython implementation, then turned around and wrote mine from scratch, and then checked that they had the same behavior with various inputs. I definitely didn't copy any part of the code, nor did I copy the whole thing and modify it. I think this counts as an "independent", rather than "derivative" work, and so can be fairly licensed under the MIT license (having read [this](https://copyleft.org/guide/comprehensive-gpl-guidech5.html)).

The CPython code in question can be found [here](https://github.com/python/cpython/blob/3ca9f50f96cfa5c1b7aa56639042531b57f07fbb/Modules/binascii.c#L436) in case anyone wants to compare.

#### RFC
I'm confused. Are you asking that I modify the docs to quote some part of the RFC? Or are you just asking me to quote it here for reviewers to see? Assuming the latter, the relevant part of the RFC is [Section 2.3](https://tools.ietf.org/html/rfc3548.html#section-2.3), which reads in part (emphasis mine):

>    Implementations MUST reject the encoding if it contains characters
>    outside the base alphabet when interpreting base encoded data, **unless
>    the specification referring to this document explicitly states
>    otherwise**.

Because our documentation explicitly states that we ignore non-alphabet characters, I think the we're still compliant.

#### Eliminating common subexpressions
> I wonder if same (more) effect could have been achieved by eliminating common subexpressions

If you're referring to eliminating common subexpressions in the original code to save space, then I kind of doubt it. I imagine that the compiler would do that for us if it made sense (i.e. the dereference will happen once and the result held in a register). I tried performing the same optimization by hand with `in[i]` in the for loop of my version. That didn't change code size at all, which I think supports that theory. On the other hand, I maybe tend to have too much faith in compiler optimizations. Also, I'm developing on a CISC machine, where I suppose subscripted dereferences might impose less of a penalty due to a richer instruction set.

I did not look at any assembly to verify any of this. If you think this warrants further investigation, or if I misunderstood your point, let me know.

**Comment by alex-robbins** [comment]
Just noticed that I missed a pair of curly braces where invalid padding is detected. Rebased and force pushed.

**Comment by pfalcon** [comment]
>  "heavily based" may have been misleading. What I meant was that my implementation was designed to have the same behavior as CPython's.

Thanks for clarifying, and please use more explicit language in the future, because if you submit a patch, and say that it's "heavily based", it's fair to assume that the code in the patch is heavily based. (Then someone analyzing licensing cleanness of MicroPython would conclude that due review diligence is not applied when accepting patches into the codebase).

> I certainly didn't copy and paste anything.

Thanks for confirming this.

**Comment by pfalcon** [comment]
> RFC
> I'm confused.

I was confused either. You added following to the docs:

> Decode base64-encoded data, ignoring invalid characters in the input.
> +   Conforms to `RFC 3548 <https://tools.ietf.org/html/rfc3548.html>`_

I doubted if ignoring invalid characters is allowed, far less required, by the RFC. That goes against well-known fail-fast approach in computing and common spirit of RFCs.

> Because our documentation explicitly states that we ignore non-alphabet characters, I think the we're still compliant.

I'm not sure to what you refer as "our documentation". In my list, that would be Python documentation, as we implement Python. There's no independent Python spec, instead CPython docs serve that purpose. And behold, nothing like that (ignoring invalid chars) is specified: https://docs.python.org/3/library/binascii.html#binascii.a2b_base64

> Convert a block of base64 data back to binary and return the binary data. More than one line may be passed at a time.

Even more interesting, b2a_base64() is marked as "The output of this function conforms to RFC 3548", while nothing like that is said for a2b_base64(), from which one might imply that CPython's a2b_base64() is not compliant with RFC 3548 (e.g. because it fails to validate input data).

However, the mystery is solved by the RFC section you mention, https://tools.ietf.org/html/rfc3548.html#section-2.3 , in the unquoted by you content:

> Such specifications may, as MIME does, instead state that   characters outside the base encoding alphabet should simply be   ignored when interpreting data ("be liberal in what you accept").

So, the reason why CPython's a2b_base64() ignores invalid chars is because MIME RFC says to do that.

Summing up:

1. This patch decreases code size, which is a boon in itself, and gets +0.9 from me.
2. We figured that *ignoring* RFC3548's suggestion of rejecting invalid chars is not wart or bug in CPython, but driven by compliance with MIME RFC. However, then suggestion:

a) either use the same trick as CPython's a2b_base64() and don't tell anything about RFC3548, or
b) dig out MIME RFC, verify it's wording, and link to it (in a2b_base64()) description instead.

Besides, there's now a conflict in the patch.

With a doc fix, solid +1 from me, thanks in advance.

File: extmod/modubinascii.c [hunk 1]
```diff
@@ -105,54 +105,64 @@ mp_obj_t mod_binascii_unhexlify(mp_obj_t data) {
...
+        nbits += 6;
+
+        if (nbits >= 8) {
+            nbits -= 8;
+            out[vstr.len++] = (shift >> nbits) & 0xFF;
         }
     }
+
+    if (nbits) {
+        mp_raise_ValueError("Incorrect padding");
```
Comment (dpgeorge) [inline extmod/modubinascii.c]: Please use lower case for the start of the message.

Comment (alex-robbins) [inline extmod/modubinascii.c]: Fixed (rebased, force pushed).

Maybe that's worth a note in `CODECONVENTIONS.md`?

**Comment by pfalcon** [comment]
> Maybe that's worth a note in CODECONVENTIONS.md?

Patches welcome, but in general, we can't specify everything. CODECONVENTIONS, etc. are there to emphasize uPy's own features or differences from CPython. All other things are like in it:

~~~
$ python3
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 1/0
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: division by zero
>>> 423rfcd
  File "<stdin>", line 1
    423rfcd
          ^
SyntaxError: invalid syntax
~~~

**Comment by pfalcon** [comment]
Updated docs to refer to MIME RFC as the normative for decoding, as suggested above, and merged. Thanks!

**Comment by alex-robbins** [comment]
Thank you.
